torch/nn/modules/module.py

```diff
@@ -193,13 +193,38 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_use_data(tensor, tensor_applied):
+            if torch._has_same_tensorimpl_type(tensor, tensor_applied):
+                # If the new tensor has the same TensorImpl type as the existing tensor,
+                # then we take `torch.__future__.get_overwrite_module_params_on_conversion()`
+                # into account.
+                return not torch.__future__.get_overwrite_module_params_on_conversion()
+            else:
+                # If the new tensor has a different TensorImpl type, we always overwrite
+                # the existing tensor.
+                return False
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                # Tensors stored in modules are graph leaves, and we don't want to
+                # track autograd history of `param_applied`, so we have to use
+                # `with torch.no_grad():`
+                with torch.no_grad():
+                    param_applied = fn(param)
```

Comment:
I tested changing `fn(param.data)` to `fn(param)` and running our test suite, and all tests passed: https://github.com/pytorch/pytorch/pull/21865. I also audited pytorch/tutorials and pytorch/examples repos, and can confirm that we are not using `_apply()` in those example use cases.

Reply:
how about stacking this so #21865 goes in first?  That will make it easier to bisect if that caused a problem.

Reply:
Sounds great - https://github.com/pytorch/pytorch/pull/21865 is ready for review

Reply:
I will rebase this PR after https://github.com/pytorch/pytorch/pull/21865 is merged.